Fix D2 check task

https://github.com/jonapoul/actual-android/actions/runs/18476621130/job/52642453679?pr=572

Incorrect path, one level out:

```
> Task :app:android:checkD2Chart FAILED
> Task :modules:budget:sync:vm:testDebugUnitTest
> Task :modules:budget:transactions:vm:testDebugUnitTest
gradle/actions: Writing build results to /home/runner/work/_temp/.gradle-actions/build-results/__run_4-1760384956117.json
OpenJDK 64-Bit Server VM warning: Sharing is only supported for boot loader classes because bootstrap classpath has been appended
OpenJDK 64-Bit Server VM warning: Sharing is only supported for boot loader classes because bootstrap classpath has been appended

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:android:checkD2Chart'.
> Chart file needs regenerating
    Generated chart file differs from the golden badging file!
      --- ...@../../../../atlas/classes.d2
      +++ ...@../../../atlas/classes.d2
          about_data: :about:data { class: module-Multiplatform }
          about_ui: :about:ui { class: module-UI }
          about_vm: :about:vm { class: module-ViewModel }
```

--- a/atlas-d2/api/atlas-d2.api
+++ b/atlas-d2/api/atlas-d2.api
@@ -28,6 +28,10 @@ public final class atlas/d2/D2AtlasPlugin : atlas/core/AtlasPlugin {
 }
 
 public abstract interface class atlas/d2/D2DagreSpec : atlas/core/PropertiesSpec {
+	public abstract fun getEdgeSep ()Ljava/lang/Integer;
+	public abstract fun getNodeSep ()Ljava/lang/Integer;
+	public abstract fun setEdgeSep (Ljava/lang/Integer;)V
+	public abstract fun setNodeSep (Ljava/lang/Integer;)V
 }
 
 public final class atlas/d2/D2DagreSpec$DefaultImpls {
@@ -464,11 +468,11 @@ public abstract class atlas/d2/tasks/ExecD2 : org/gradle/api/DefaultTask, atlas/
 public abstract class atlas/d2/tasks/WriteD2Chart : org/gradle/api/DefaultTask, atlas/core/tasks/AtlasGenerationTask, atlas/core/tasks/TaskWithOutputFile {
 	public fun <init> ()V
 	public fun execute ()V
-	public abstract fun getGlobalFile ()Lorg/gradle/api/file/RegularFileProperty;
 	public abstract fun getGroupModules ()Lorg/gradle/api/provider/Property;
 	public abstract fun getLinksFile ()Lorg/gradle/api/file/RegularFileProperty;
 	public abstract fun getModuleTypesFile ()Lorg/gradle/api/file/RegularFileProperty;
 	public abstract fun getOutputFile ()Lorg/gradle/api/file/RegularFileProperty;
+	public abstract fun getPathToClassesFile ()Lorg/gradle/api/provider/Property;
 	public abstract fun getReplacements ()Lorg/gradle/api/provider/SetProperty;
 	public abstract fun getThisPath ()Lorg/gradle/api/provider/Property;
 }

--- a/atlas-d2/src/main/kotlin/atlas/d2/D2AtlasPlugin.kt
+++ b/atlas-d2/src/main/kotlin/atlas/d2/D2AtlasPlugin.kt
@@ -51,16 +51,25 @@ public class D2AtlasPlugin : AtlasPlugin() {
   override fun Project.registerChildTasks() {
     val d2Spec = d2Extension.d2
 
+    // need to use the same pathToClassesFile string for real and dummy tasks, otherwise the check operation might
+    // fail if the module and the build directory have different relative paths.
+    val writeD2Classes = WriteD2Classes.get(rootProject)
+    val classesFile = writeD2Classes.map { it.outputFile.get() }
+    val outputFile = outputFile(Chart, d2Spec.fileExtension.get())
+    val pathToClassesFile = classesFile.map { it.asFile.relativeTo(outputFile.parentFile).path }
+
     val chartTask = WriteD2Chart.real(
       target = project,
       extension = extension,
-      outputFile = outputFile(Chart, d2Spec.fileExtension.get()),
+      outputFile = outputFile,
+      pathToClassesFile = pathToClassesFile,
     )
 
     val dummyChartTask = WriteD2Chart.dummy(
       target = project,
       extension = extension,
       outputFile = atlasBuildDirectory.get().file("chart-temp.d2").asFile,
+      pathToClassesFile = pathToClassesFile,
     )
 
     CheckFileDiff.register(

--- a/atlas-d2/src/main/kotlin/atlas/d2/internal/D2Writer.kt
+++ b/atlas-d2/src/main/kotlin/atlas/d2/internal/D2Writer.kt
@@ -25,7 +25,7 @@ public class D2Writer(
   override val replacements: Set<Replacement>,
   override val thisPath: String,
   override val groupModules: Boolean,
-  private val globalRelativePath: String,
+  private val pathToClassesFile: String,
 ) : ChartWriter() {
   private var subgraphNestingLevel = 0
 
@@ -37,7 +37,7 @@ public class D2Writer(
   }
 
   private fun IndentedStringBuilder.appendImports() {
-    appendLine("...@$globalRelativePath")
+    appendLine("...@$pathToClassesFile")
   }
 
   override fun IndentedStringBuilder.appendSubgraphHeader(graph: Subgraph) {

--- a/atlas-d2/src/main/kotlin/atlas/d2/tasks/WriteD2Chart.kt
+++ b/atlas-d2/src/main/kotlin/atlas/d2/tasks/WriteD2Chart.kt
@@ -21,6 +21,7 @@ import org.gradle.api.DefaultTask
 import org.gradle.api.Project
 import org.gradle.api.file.RegularFileProperty
 import org.gradle.api.provider.Property
+import org.gradle.api.provider.Provider
 import org.gradle.api.provider.SetProperty
 import org.gradle.api.tasks.CacheableTask
 import org.gradle.api.tasks.Input
@@ -36,9 +37,9 @@ import java.io.File
 @CacheableTask
 public abstract class WriteD2Chart : DefaultTask(), TaskWithOutputFile, AtlasGenerationTask {
   // Files
-  @get:[PathSensitive(RELATIVE) InputFile] public abstract val globalFile: RegularFileProperty
   @get:[PathSensitive(RELATIVE) InputFile] public abstract val linksFile: RegularFileProperty
   @get:[PathSensitive(RELATIVE) InputFile] public abstract val moduleTypesFile: RegularFileProperty
+  @get:Input public abstract val pathToClassesFile: Property<String>
   @get:OutputFile abstract override val outputFile: RegularFileProperty
 
   // General
@@ -53,10 +54,10 @@ public abstract class WriteD2Chart : DefaultTask(), TaskWithOutputFile, AtlasGen
 
   @TaskAction
   public open fun execute() {
-    val globalFile = globalFile.get().asFile
     val linksFile = linksFile.get().asFile
     val moduleTypesFile = moduleTypesFile.get().asFile
     val outputFile = outputFile.get().asFile
+    val pathToClassesFile = pathToClassesFile.get()
 
     val writer = D2Writer(
       typedModules = readModuleTypes(moduleTypesFile),
@@ -64,7 +65,7 @@ public abstract class WriteD2Chart : DefaultTask(), TaskWithOutputFile, AtlasGen
       replacements = replacements.get(),
       thisPath = thisPath.get(),
       groupModules = groupModules.get(),
-      globalRelativePath = globalFile.relativeTo(outputFile.parentFile).path,
+      pathToClassesFile = pathToClassesFile,
     )
 
     outputFile.writeText(writer())
@@ -79,28 +80,30 @@ public abstract class WriteD2Chart : DefaultTask(), TaskWithOutputFile, AtlasGen
       target: Project,
       extension: AtlasExtension,
       outputFile: File,
-    ) = register<WriteD2Chart>(target, extension, outputFile)
+      pathToClassesFile: Provider<String>,
+    ) = register<WriteD2Chart>(target, extension, outputFile, pathToClassesFile)
 
     internal fun dummy(
       target: Project,
       extension: AtlasExtension,
       outputFile: File,
-    ) = register<WriteD2ChartDummy>(target, extension, outputFile)
+      pathToClassesFile: Provider<String>,
+    ) = register<WriteD2ChartDummy>(target, extension, outputFile, pathToClassesFile)
 
-    internal inline fun <reified T : WriteD2Chart> register(
+    private inline fun <reified T : WriteD2Chart> register(
       target: Project,
       extension: AtlasExtension,
       outputFile: File,
+      pathToClassesFile: Provider<String>,
     ): TaskProvider<T> = with(target) {
       val collateModuleTypes = CollateModuleTypes.get(rootProject)
       val calculateProjectTree = WriteModuleTree.get(target)
-      val writeD2Classes = WriteD2Classes.get(rootProject)
       val name = "write${T::class.qualifier}D2Chart"
       val writeChart = tasks.register(name, T::class.java) { task ->
-        task.globalFile.convention(writeD2Classes.map { it.outputFile.get() })
         task.linksFile.convention(calculateProjectTree.map { it.outputFile.get() })
         task.moduleTypesFile.convention(collateModuleTypes.map { it.outputFile.get() })
         task.outputFile.set(outputFile)
+        task.pathToClassesFile.convention(pathToClassesFile)
         task.thisPath.convention(target.path)
       }
 

--- a/atlas-test/src/test/kotlin/atlas/d2/Factories.kt
+++ b/atlas-test/src/test/kotlin/atlas/d2/Factories.kt
@@ -20,5 +20,5 @@ internal fun d2Writer(
   replacements = replacements,
   thisPath = thisPath,
   groupModules = groupModules,
-  globalRelativePath = classesRelativePath,
+  pathToClassesFile = classesRelativePath,
 )

--- a/atlas-test/src/test/kotlin/atlas/d2/tasks/WriteD2ChartTest.kt
+++ b/atlas-test/src/test/kotlin/atlas/d2/tasks/WriteD2ChartTest.kt
@@ -5,11 +5,15 @@
 package atlas.d2.tasks
 
 import assertk.assertThat
+import assertk.assertions.exists
+import atlas.d2.RequiresD2
 import atlas.test.ScenarioTest
 import atlas.test.allTasksSuccessful
 import atlas.test.contentEquals
 import atlas.test.runTask
 import atlas.test.scenarios.D2Basic
+import atlas.test.scenarios.D2NestedModules
+import atlas.test.taskWasSuccessful
 import kotlin.test.Test
 
 internal class WriteD2ChartTest : ScenarioTest() {
@@ -67,5 +71,32 @@ internal class WriteD2ChartTest : ScenarioTest() {
         }
       """.trimIndent(),
     )
+  }
+
+  @Test
+  @RequiresD2
+  fun `Write correct classes file path for nested modules`() = runScenario(D2NestedModules) {
+    // when
+    val result = runTask("atlasGenerate").build()
+
+    // then
+    assertThat(result).allTasksSuccessful()
+
+    // and the file was generated
+    val chartFile = resolve("path/to/my/module/atlas/chart.d2")
+    assertThat(chartFile.exists())
+    assertThat(resolve("atlas/classes.d2")).exists()
+    assertThat(chartFile).contentEquals(
+      """
+        ...@../../../../../atlas/classes.d2
+        path_to_my_module: :path:to:my:module
+      """.trimIndent(),
+    )
+
+    // when we check
+    val checkResult = runTask("check").build()
+
+    // then
+    assertThat(checkResult).taskWasSuccessful(":path:to:my:module:checkD2Chart")
   }
 }

--- a/atlas-test/src/test/kotlin/atlas/test/Assertions.kt
+++ b/atlas-test/src/test/kotlin/atlas/test/Assertions.kt
@@ -39,7 +39,9 @@ internal fun Assert<BuildTask?>.hadResult(expected: TaskOutcome) = given { resul
 internal fun Assert<File>.contentEquals(expected: String) = given { file ->
   val contents = file.readText().removeSuffix("\n")
   if (contents == expected) return@given
-  fail("Failed contentEquals: expected='$expected', actual='$contents'")
+  expected(
+    "Unequal strings between expected{${expected.length}} and actual{${contents.length}}:\n" + diff(expected, contents),
+  )
 }
 
 // Copied from assertk repo but they haven't published in ages

--- a/atlas-test/src/test/kotlin/atlas/test/scenarios/D2NestedModules.kt
+++ b/atlas-test/src/test/kotlin/atlas/test/scenarios/D2NestedModules.kt
@@ -1,0 +1,22 @@
+/**
+ * Copyright © 2025 Jon Poulton
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package atlas.test.scenarios
+
+import atlas.test.D2Scenario
+import atlas.test.KOTLIN_VERSION
+import atlas.test.kotlinJvmBuildScript
+
+internal object D2NestedModules : D2Scenario {
+  override val rootBuildFile = """
+    plugins {
+      kotlin("jvm") version "$KOTLIN_VERSION" apply false
+      id("$pluginId")
+    }
+  """.trimIndent()
+
+  override val submoduleBuildFiles = mapOf(
+    "path:to:my:module" to kotlinJvmBuildScript,
+  )
+}

--- a/gradle.properties
+++ b/gradle.properties
@@ -21,7 +21,7 @@ atlas.minimumGradleVersion=8.13
 GROUP=dev.jonpoulton.atlas
 POM_NAME=Atlas Gradle Plugin
 POM_DESCRIPTION=A set of Gradle plugins for generating diagrams of modular dependencies
-VERSION_NAME=0.2.0-SNAPSHOT
+VERSION_NAME=0.1.1
 
 POM_DEVELOPER_ID=jonapoul
 POM_DEVELOPER_NAME=Jon Poulton

